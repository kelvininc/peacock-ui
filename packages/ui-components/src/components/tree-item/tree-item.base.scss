@import '../../assets/styles/globals';

:host {
	/**
	 * @prop --node-width: Tree Node height.
	 * @prop --node-height: Tree Node height.
	 * @prop --node-gap: Gap between child nodes in px.
	 * @prop --children-offset: Offset of the child nodes in px.

	 * @prop --connector-lines-color: Line color of children's connectors.

	 * @prop --background-color-default: Background color when state is default.
	 * @prop --background-color-focused: Background color when state is focused.
	 * @prop --background-color-disabled: Background color when state is disabled.
	 * @prop --background-color-highlighted: Background color when state is highlighted.

	 * @prop --border-color-default: Border color when state is default.
	 * @prop --border-color-disabled: Border color when state is disabled.
	 * @prop --border-color-selected: Border color when state is selected.
	 * @prop --border-color-highlighted: Border color when state is highlighted.

	 * @prop --icon-color-default: Node icon color when node state is default.
	 * @prop --icon-color-disabled: Node icon color when node state is disabled.
	 * @prop --icon-color-selected: Node icon color when node state is selected.
	 * @prop --icon-color-no-filled: Node icon color when node state is no-filled.

	 * @prop --title-color-default: Node title color when node state is default.
	 * @prop --title-color-disabled: Node title color when node state is disabled.
	 * @prop --title-color-selected: Node title color when node state is selected.
	 * @prop --title-color-no-filled: Node title color when node state is not filled.

	 * @prop --sub-title-color-default: Node sub-title color when node state is default.
	 * @prop --sub-title-color-disabled: Node sub-title color when node state is disabled.
	 * @prop --sub-title-color-no-filled: Node sub-title color when node state is not filled.
	 */

	--node-width: 300px;
	--node-height: 40px;
	--node-gap: #{$spacing-2x};
	--children-offset: #{$spacing-3x};

	// Private usage
	/* stylelint-disable-next-line length-zero-no-unit */
	--private-left-shift: 0px;
	--private-node-gap: 0;
	--private-vertical-line-height: 0;

	.node-container {
		position: relative;

		// Vertical Lines
		&::after {
			content: '';
			position: absolute;
			border-left: 1px dashed var(--connector-lines-color);
			height: calc(var(--private-vertical-line-height) + var(--private-node-gap));
			width: 0;
			top: calc(-1 * var(--private-node-gap));
		}
	}

	.children {
		margin-left: calc($spacing-5x + var(--private-left-shift));

		::slotted(kv-tree-item) {
			--private-left-shift: #{var(--children-offset)};
			--private-node-gap: #{var(--node-gap)};
			--private-vertical-line-height: 100%;
		}

		::slotted(kv-tree-item:last-child) {
			--private-vertical-line-height: calc(var(--node-height) / 2 +4px);
		}
	}

	.node-loading {
		display: flex;
		width: 100%;
		height: 100%;
	}

	.node-wrapper {
		position: relative;
		display: flex;
		width: var(--node-width);
		height: var(--node-height);
		margin: var(--private-node-gap) 0 0 var(--private-left-shift);
		align-items: center;
		box-sizing: border-box;
		padding: $spacing-2x 0;
		background-color: var(--background-color-default);
		border: 1px solid var(--border-color-default);
		border-radius: 2px;

		// Horizontal Lines
		&::before {
			content: '';
			position: absolute;
			left: calc(-1 * var(--private-left-shift));
			width: calc(var(--private-left-shift) - 2px);
			border-top: 1px dashed var(--connector-lines-color);
		}

		&:hover,
		&.selected:hover,
		&.highlighted:hover {
			background-color: var(--background-color-focused);
		}

		&.disabled {
			background-color: var(--background-color-disabled);
			border-color: var(--border-color-disabled);
		}

		&.selected {
			background-color: var(--background-color-focused);
			border-color: var(--border-color-selected);

			/* stylelint-disable-next-line selector-max-compound-selectors*/
			.expander-arrow kv-icon {
				--icon-color: var(--icon-color-selected);
			}
		}

		&.highlighted {
			background-color: var(--background-color-highlighted);
			border-color: var(--border-color-highlighted);
		}

		&.loading {
			padding: 0;
			background-color: unset;
			border: unset;
		}
	}

	.expander-arrow {
		cursor: pointer;
		padding-left: $spacing-2x;

		kv-icon {
			--icon-color: var(--icon-color-default);
			display: flex;
			align-items: center;
		}
	}

	.node-content-wrapper {
		flex: 1;
		min-width: 0;
		height: 100%;
		display: flex;
		align-items: center;
		padding: 0 $spacing-2x;
		border-radius: 0;
		cursor: pointer;

		.node-icon {
			position: relative;
			width: 24px;
			height: 24px;
		}

		.main-icon {
			--icon-color: var(--icon-color-default);

			display: inline-flex;
			align-items: center;
			justify-content: center;
		}

		.state-icon {
			--icon-color: #{kv-color('neutral-1')};
			position: absolute;
			right: 0;
			bottom: -5px;

			&.info {
				--icon-color: #{kv-color('info')};
			}

			&.warning {
				--icon-color: #{kv-color('warning')};
			}

			&.success {
				--icon-color: #{kv-color('success')};
			}

			&.error {
				--icon-color: #{kv-color('error')};
			}
		}

		.labels {
			@include kv-font-label-small-regular();
			flex: 1;
			min-width: 0;
			margin-left: $spacing-2x;
			display: inline-flex;
			flex-direction: column;
			color: var(--title-color-default);
		}

		.title {
			@include ellipsis();
		}

		.sub-title {
			@include kv-font-label-xsmall-uppercase-regular();
			@include ellipsis();
			color: var(--sub-title-color-default);
		}

		&.no-filled {
			.main-icon {
				--icon-color: var(--icon-color-no-filled);
			}

			.state-icon {
				opacity: 0.5;
			}

			.title {
				color: var(--title-color-no-filled);
			}

			.sub-title {
				color: var(--sub-title-color-no-filled);
			}
		}

		&.disabled {
			cursor: not-allowed;

			.main-icon {
				--icon-color: var(--icon-color-disabled);
			}

			.state-icon {
				opacity: 0.5;
			}

			.title {
				color: var(--title-color-disabled);
			}

			.sub-title {
				color: var(--sub-title-color-disabled);
			}

			.alarm-bubble {
				opacity: 0.5;
			}
		}

		&.selected {
			.main-icon {
				--icon-color: var(--icon-color-selected);
			}

			.title {
				@include kv-font-label-small-semibold();
				color: var(--title-color-selected);
			}
		}
	}

	.alarm-bubble {
		width: 16px;
		height: 16px;
		background-color: kv-color('neutral-5');
		border-radius: 50%;
		text-align: center;
		line-height: 15px;

		span {
			@include kv-font-label-xsmall-uppercase-regular();
			vertical-align: middle;
			color: kv-color('neutral-0');

			font-size: 8px;
		}

		&.info {
			background-color: kv-color('info');
		}

		&.warning {
			background-color: kv-color('warning');
		}

		&.success {
			background-color: kv-color('success');
		}

		&.error {
			background-color: kv-color('error');
		}
	}
}
