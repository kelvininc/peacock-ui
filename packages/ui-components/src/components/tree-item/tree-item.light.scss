@import '../../assets/styles/globals';
@import './tree-item.base';

:host {
	--connector-lines-color: #{kv-color('primary')};

	--background-color-default: #{kv-color('neutral-0')};
	--background-color-focused: #{kv-color('primary', 'base', 0.1)};
	--background-color-disabled: #{kv-color('neutral-1')};
	--background-color-highlighted: #{kv-color('primary', 'base', 0.1)};

	--border-color-default: #{kv-color('neutral-3')};
	--border-color-disabled: #{kv-color('neutral-3')};
	--border-color-selected: #{kv-color('neutral-3')};
	--border-color-highlighted: #{kv-color('neutral-3')};

	--icon-color-default: #{kv-color('neutral-3')};
	--icon-color-disabled: #{kv-color('neutral-3')};
	--icon-color-selected: #{kv-color('primary')};
	--icon-color-no-filled: #{kv-color('neutral-3')};

	--title-color-default: #{kv-color('neutral-9')};
	--title-color-disabled: #{kv-color('neutral-5')};
	--title-color-selected: #{kv-color('neutral-9')};
	--title-color-no-filled: #{kv-color('neutral-5')};

	--sub-title-color-default: #{kv-color('neutral-5')};
	--sub-title-color-disabled: #{kv-color('neutral-5')};
	--sub-title-color-no-filled: #{kv-color('neutral-5')};
}

.node-loading {
	@include ghost-animation('neutral-2');
}
