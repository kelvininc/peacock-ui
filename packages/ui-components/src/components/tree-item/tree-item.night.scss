@import '../../assets/styles/globals';
@import './tree-item.base';

:host {
	--connector-lines-color: #{kv-color('neutral-5')};

	--background-color-default: #{kv-color('neutral-6')};
	--background-color-focused: #{kv-color('neutral-5')};
	--background-color-disabled: #{kv-color('neutral-7')};
	--background-color-highlighted: #{kv-color('neutral-6')};

	--border-color-default: #{kv-color('neutral-5')};
	--border-color-disabled: #{kv-color('neutral-6')};
	--border-color-selected: #{kv-color('neutral-0')};
	--border-color-highlighted: #{kv-color('neutral-4')};

	--icon-color-default: #{kv-color('neutral-2')};
	--icon-color-disabled: #{kv-color('neutral-5')};
	--icon-color-selected: #{kv-color('neutral-0')};
	--icon-color-no-filled: #{kv-color('neutral-4')};

	--title-color-default: #{kv-color('neutral-2')};
	--title-color-disabled: #{kv-color('neutral-5')};
	--title-color-selected: #{kv-color('neutral-0')};
	--title-color-no-filled: #{kv-color('neutral-4')};

	--sub-title-color-default: #{kv-color('neutral-4')};
	--sub-title-color-disabled: #{kv-color('neutral-5')};
	--sub-title-color-no-filled: #{kv-color('neutral-4')};
}

.node-loading {
	@include ghost-animation('neutral-7');
}
