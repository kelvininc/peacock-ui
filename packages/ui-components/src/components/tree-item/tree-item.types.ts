export enum ETreeItemState {
	None = 'none',
	Info = 'info',
	Warning = 'warning',
	Error = 'error',
	Success = 'success'
}
