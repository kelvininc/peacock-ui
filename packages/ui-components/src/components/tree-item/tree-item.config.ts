import { EIconName } from '../icon/icon.types';
import { ETreeItemState } from './tree-item.types';

export const STATE_ICONS: { [key in ETreeItemState]: EIconName } = {
	[ETreeItemState.Success]: EIconName.Success,
	[ETreeItemState.Error]: EIconName.Error,
	[ETreeItemState.Warning]: EIconName.Warning,
	[ETreeItemState.Info]: EIconName.Info,
	[ETreeItemState.None]: null
};
