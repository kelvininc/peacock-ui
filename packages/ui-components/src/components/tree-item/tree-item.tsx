import { Component, Element, Event, EventEmitter, Fragment, Host, Listen, Prop, h } from '@stencil/core';
import { EAnchorTarget, IAnchor } from '../../utils/types';
import { EIconName, EOtherIconName } from '../icon/icon.types';
import { isEmpty, isNumber, throttle } from 'lodash-es';

import { DEFAULT_THROTTLE_WAIT } from '../../config';
import { EBadgeState } from '../badge/badge.types';
import { ETreeItemState } from './tree-item.types';
import { STATE_ICONS } from './tree-item.config';

/**
 * @slot child-slot - Content is placed in the child subgroup and can be expanded and collapsed.
 * @part children - The children container.
 */
@Component({
	tag: 'kv-tree-item',
	styleUrls: {
		night: 'tree-item.night.scss',
		light: 'tree-item.light.scss'
	},
	shadow: true
})
export class KvTreeItem implements IAnchor {
	/** (optional) Defines the title of the tree item.*/
	@Prop({ reflect: true }) label?: string;
	/** (optional) Defines the sub-title of the tree item, displayed under the title.*/
	@Prop({ reflect: true }) additionalLabel?: string;
	/** (optional) Defines the placeholder of the tree item, displayed when title is not filled.*/
	@Prop({ reflect: true }) placeholder?: string;
	/** (optional) Defines the icon of the tree item. If set, an icon will be displayed before the label.*/
	@Prop({ reflect: true }) icon?: EIconName | EOtherIconName;
	/** (optional) Defines the state of the icon.*/
	@Prop({ reflect: true }) iconState?: ETreeItemState;
	/** (optional) Defines the counter info of the tree item. If set, an badge will be displayed in the end of tree item.*/
	@Prop({ reflect: true }) counter?: number;
	/** (optional) Defines the state of the counter.*/
	@Prop({ reflect: true }) counterState?: EBadgeState;
	/** (optional) Defines whether the tree node has children, even if currently no other tree nodes are slotted inside.
	 * This property is useful for showing big tree structures where not all nodes are initially loaded due to performance reasons.
	 * Set this to <code>true</code> for nodes you intend to load lazily, when the user clicks the expand button.
	 * It is not necessary to set this property otherwise. If a tree item has children, the expand button will be displayed anyway.
	 */
	@Prop({ reflect: true }) hasChildren? = false;
	/** (optional) Defines whether the tree node is expanded or collapsed. Only has visual effect for tree nodes with children.*/
	@Prop({ reflect: true }) expanded? = false;
	/** (optional) Defines whether the tree node is disabled.*/
	@Prop({ reflect: true }) disabled? = false;
	/** (optional) Defines whether the tree node is selected by the user.*/
	@Prop({ reflect: true }) selected? = false;
	/** (optional) Defines whether the tree node is highlighted.*/
	@Prop({ reflect: true }) highlighted? = false;
	/** (optional) Defines whether the tree node is spotlight.*/
	@Prop({ reflect: true }) spotlighted? = false;
	/** (optional) Defines whether the tree node is loading. */
	@Prop({ reflect: true }) loading? = false;
	/** (optional) Defines if the item click event should prevent default behaviour. */
	@Prop({ reflect: true }) preventDefault? = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) href?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) target?: EAnchorTarget;
	/** @inheritdoc */
	@Prop({ reflect: true }) download?: string;

	/** Emitted when the expand toggle is clicked */
	@Event() toggleExpand: EventEmitter<MouseEvent>;
	/** Emitted when the tree item is clicked */
	@Event() itemClick: EventEmitter<MouseEvent>;

	/** Don't propagate the children click event */
	@Listen('itemClick')
	itemClickHandler(event: MouseEvent) {
		event.stopPropagation();
	}

	/** Don't propagate the children toggle event */
	@Listen('toggleExpand')
	toggleExpandHandler(event: MouseEvent) {
		event.stopPropagation();
	}

	// Reference to self
	@Element() el: HTMLKvTreeItemElement;

	private toggleClickThrottler: (e: MouseEvent) => void;
	private itemClickThrottler: (e: MouseEvent) => void;

	private get hasChildrenSlot() {
		return !isEmpty(this.el.querySelector('[slot="child-slot"]'));
	}

	get requiresToggleButton() {
		return this.hasChildren || this.hasChildrenSlot;
	}

	connectedCallback() {
		this.toggleClickThrottler = throttle((event: MouseEvent) => this.toggleExpand.emit(event), DEFAULT_THROTTLE_WAIT);
		this.itemClickThrottler = throttle((event: MouseEvent) => this.itemClick.emit(event), DEFAULT_THROTTLE_WAIT);
	}

	private onItemClick(event: MouseEvent) {
		if (this.preventDefault) {
			event.preventDefault();
		}

		this.itemClickThrottler(event);
	}

	render() {
		return (
			<Host>
				<div class="node-container">
					<div
						class={{
							'node-wrapper': true,
							'disabled': this.disabled,
							'selected': this.selected,
							'highlighted': this.highlighted,
							'spotlight': this.spotlighted,
							'loading': this.loading
						}}
					>
						{!this.loading && (
							<Fragment>
								{this.requiresToggleButton && (
									<div class="expander-arrow" onClick={this.toggleClickThrottler}>
										<kv-icon name={this.expanded ? EIconName.ArrowDropDown : EIconName.ArrowRight} customClass="pk-grey3 icon-24" />
									</div>
								)}

								<a
									class={{
										'node-content-wrapper': true,
										'disabled': this.disabled,
										'selected': this.selected,
										'no-filled': isEmpty(this.label) && !isEmpty(this.placeholder)
									}}
									download={this.download}
									href={this.href}
									target={this.target}
									onClick={!this.disabled && this.onItemClick.bind(this)}
								>
									{this.icon && (
										<div class="node-icon">
											<kv-icon name={this.icon} customClass="icon-24" class="main-icon" exportparts="icon"></kv-icon>
											{this.iconState && (
												<kv-icon name={STATE_ICONS[this.iconState]} customClass="icon-12" class={{ 'state-icon': true, [this.iconState]: true }} />
											)}
										</div>
									)}

									{(this.label || this.placeholder) && (
										<div class="labels">
											<div class="title">{this.label || this.placeholder}</div>
											{this.additionalLabel && <div class="sub-title">{this.additionalLabel}</div>}
										</div>
									)}

									<div class="right-indicators">
										{isNumber(this.counter) && this.counter >= 0 && (
											<div class="alarm-bubble">
												<kv-badge state={this.counterState}>{this.counter > 100 ? '+99' : this.counter}</kv-badge>
											</div>
										)}
									</div>
								</a>
							</Fragment>
						)}
						{this.loading && <div class="node-loading"></div>}
					</div>
					<div part="children" class="children" style={{ display: this.expanded ? 'block' : 'none' }}>
						<slot name="child-slot"></slot>
					</div>
				</div>
			</Host>
		);
	}
}
