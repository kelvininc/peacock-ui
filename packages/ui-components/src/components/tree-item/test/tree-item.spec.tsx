import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { KvTreeItem } from '../tree-item';

describe('Tree Item (unit tests)', () => {
	let page: SpecPage;
	let component: KvTreeItem;

	describe('when initialized with required props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTreeItem],
				html: `<kv-tree-item label="Node Title"></kv-tree-item>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the label prop value', () => {
			expect(component.label).toEqual('Node Title');
		});
	});

	describe('when initialized with disabled attribute', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTreeItem],
				html: `<kv-tree-item label="Node Title" disabled></kv-tree-item>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the disabled prop value', () => {
			expect(component.disabled).toBeTruthy();
		});
	});

	describe('when initialized with selected attribute', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTreeItem],
				html: `<kv-tree-item label="Node Title" selected></kv-tree-item>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the selected prop value', () => {
			expect(component.selected).toBeTruthy();
		});
	});

	describe('when initialized with counter attribute', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTreeItem],
				html: `<kv-tree-item label="Node Title" counter="32"></kv-tree-item>`
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should set the counter prop value', () => {
			expect(component.counter).toEqual(32);
		});
	});

	describe('when initialized with preventDefault prop', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTreeItem],
				html: '<kv-tree-item label="Node Title" prevent-default="true"></kv-tree-item>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		describe('and onItemClick is called', () => {
			let onClickSpyEvent: MouseEvent;

			beforeEach(() => {
				onClickSpyEvent = new MouseEvent('click');
				component.onItemClick(onClickSpyEvent);
			});

			it('should prevent event', () => {
				expect(onClickSpyEvent.defaultPrevented).toBeTruthy();
			});
		});
	});
});
