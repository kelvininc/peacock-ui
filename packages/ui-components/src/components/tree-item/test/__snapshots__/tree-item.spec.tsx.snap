// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Item (unit tests) when initialized with counter attribute should match the snapshot 1`] = `
<kv-tree-item counter="32" label="Node Title">
  <mock:shadow-root>
    <div class="node-container">
      <div class="node-wrapper">
        <div class="node-content-wrapper">
          <div class="labels">
            <div class="title">
              Node Title
            </div>
          </div>
          <div class="right-indicators">
            <div class="alarm-bubble undefined">
              <span>
                32
              </span>
            </div>
          </div>
        </div>
      </div>
      <div class="children" style="display: none;">
        <slot name="child-slot"></slot>
      </div>
    </div>
  </mock:shadow-root>
</kv-tree-item>
`;

exports[`Tree Item (unit tests) when initialized with disabled attribute should match the snapshot 1`] = `
<kv-tree-item disabled="" label="Node Title">
  <mock:shadow-root>
    <div class="node-container">
      <div class="disabled node-wrapper">
        <div class="disabled node-content-wrapper">
          <div class="labels">
            <div class="title">
              Node Title
            </div>
          </div>
          <div class="right-indicators"></div>
        </div>
      </div>
      <div class="children" style="display: none;">
        <slot name="child-slot"></slot>
      </div>
    </div>
  </mock:shadow-root>
</kv-tree-item>
`;

exports[`Tree Item (unit tests) when initialized with required props should match the snapshot 1`] = `
<kv-tree-item label="Node Title">
  <mock:shadow-root>
    <div class="node-container">
      <div class="node-wrapper">
        <div class="node-content-wrapper">
          <div class="labels">
            <div class="title">
              Node Title
            </div>
          </div>
          <div class="right-indicators"></div>
        </div>
      </div>
      <div class="children" style="display: none;">
        <slot name="child-slot"></slot>
      </div>
    </div>
  </mock:shadow-root>
</kv-tree-item>
`;

exports[`Tree Item (unit tests) when initialized with selected attribute should match the snapshot 1`] = `
<kv-tree-item label="Node Title" selected="">
  <mock:shadow-root>
    <div class="node-container">
      <div class="node-wrapper selected">
        <div class="node-content-wrapper selected">
          <div class="labels">
            <div class="title">
              Node Title
            </div>
          </div>
          <div class="right-indicators"></div>
        </div>
      </div>
      <div class="children" style="display: none;">
        <slot name="child-slot"></slot>
      </div>
    </div>
  </mock:shadow-root>
</kv-tree-item>
`;
